dns/dns.go

```diff
@@ -1,35 +1,141 @@
 package dns
 
 import (
+	"bufio"
 	"errors"
 	"fmt"
 	"log"
+	"os"
+	"path"
 	"strconv"
+	"strings"
 	"sync"
+	"time"
 
 	"github.com/MarekWojt/gertdns/config"
+	"github.com/MarekWojt/gertdns/util"
+	"github.com/gookit/color"
 	"github.com/miekg/dns"
 )
 
+var domains []*domain = make([]*domain, 0)
+var saveTicker *time.Ticker = time.NewTicker(time.Second)
+var saving sync.Mutex = sync.Mutex{}
+var currentDataPath string
+
+const (
+	IPV4_FILE = "v4.csv"
+	IPV6_FILE = "v6.csv"
+)
+
 type domain struct {
-	Root  string
-	Mutv4 sync.RWMutex
-	Mutv6 sync.RWMutex
-	Ipv4  map[string]string
-	Ipv6  map[string]string
+	Root         string
+	Mutv4        sync.RWMutex
+	Mutv6        sync.RWMutex
+	Mutv4Changed sync.RWMutex
+	Mutv6Changed sync.RWMutex
+	Ipv4         map[string]string
+	Ipv4Changed  bool
+	Ipv6         map[string]string
+	Ipv6Changed  bool
 }
 
-var domains []*domain = make([]*domain, 0)
+func (currentDomain *domain) IsV4Changed() bool {
+	currentDomain.Mutv4Changed.RLock()
+	result := currentDomain.Ipv4Changed
+	currentDomain.Mutv4Changed.RUnlock()
+
+	return result
+}
+
+func (currentDomain *domain) IsV6Changed() bool {
+	currentDomain.Mutv6Changed.RLock()
+	result := currentDomain.Ipv6Changed
+	currentDomain.Mutv6Changed.RUnlock()
+
+	return result
+}
+
+func (currentDomain *domain) MarkV6Changed(changed bool) {
+	currentDomain.Mutv6Changed.Lock()
+	currentDomain.Ipv6Changed = changed
+	currentDomain.Mutv6Changed.Unlock()
+}
+
+func (currentDomain *domain) MarkV4Changed(changed bool) {
+	currentDomain.Mutv4Changed.Lock()
+	currentDomain.Ipv4Changed = changed
+	currentDomain.Mutv4Changed.Unlock()
+}
 
-func Init() {
+func (currentDomain *domain) SetV4(domain string, ipv4 string) {
+	currentDomain.Mutv4.Lock()
+	currentDomain.Ipv4[domain] = ipv4
+	currentDomain.Mutv4.Unlock()
+}
+
+func (currentDomain *domain) SetV6(domain string, ipv6 string) {
+	currentDomain.Mutv6.Lock()
+	currentDomain.Ipv6[domain] = ipv6
+	currentDomain.Mutv6.Unlock()
+}
+
+func loadFile(ty string, currentDomain *domain) {
+	if ty != IPV4_FILE && ty != IPV6_FILE {
+		panic("type passed to loadFile must be either IPV4_FILE or IPV6_FILE")
+	}
+
+	filePath := path.Join(currentDataPath, currentDomain.Root+ty)
+	f, err := os.Open(filePath)
```

Comment:
pls use `defer f.Close()` to always close the file even if some error occures during the scan logic.

Reply:
is now implemented with the latest commits